### DISCUSSION TITLE
Possibility to use all the vertical_strategy and horizontal_strategy

### DISCUSSION THREAD

**Discussion by joelpeter-c** [jsvine/pdfplumber#721] [Q&A]
Hi All, 

Looking to see, if there is a way to use all possible strategies to extract tables? I encounter tables of all possible strategy in the pdf.

Right now, I am applying all the possible `vertical_strategies` and `horizontal_strategies` on the page until I find no tables for a page and this is inturn causing slowness in analysis of the pdf page by page.

Is there a better way to do it?

Thanks in advance.

**Comment by jsvine** [comment]
Hi @joelpeter94, and thanks for your interest in this library. Off the top of my head, I can't think of an approach that would be more efficient than the one you're describing.

**Reply by joelpeter-c** [reply]
Hi @jsvine  Thank you for the reply, its an awesome library.

Please find my code below for clarity


    
    for page in pdf.pages:
        table = page.extract_table(table_settings=
        {
            "vertical_strategy": "text", 
            "horizontal_strategy": "text", 
            "min_words_vertical": 2, 
            "min_words_horizontal": 2,
            "keep_blank_chars": True
        })
        #if table is empty try horizontal_strategy =lines approach 
        if table is None:
            table = page.extract_table(table_settings=
            {
                "vertical_strategy": "text", 
                "horizontal_strategy": "lines", 
                "min_words_vertical": 2, 
                "min_words_horizontal": 2,
                "keep_blank_chars": True
            })
            # table is empty and no images (image check is done to handle charts)
            if table is None and not bool(page.images):
                table = page.extract_table(table_settings=
                {
                    "vertical_strategy": "lines", 
                    "horizontal_strategy": "lines", 
                    "min_words_vertical": 2, 
                    "min_words_horizontal": 2,
                    "keep_blank_chars": True
                })
                if table is None:
                    pagenumbers.append(page.page_number-1)

        page.flush_cache()

    print(set(pagenumbers))


I basically want to find if a table exist in a pdf page or not, but it's too slow and I am assuming it is because of the multiple approaches I try to check with. Is there something that I can tweak or things in the library that I can use to make it efficient?

**Reply by jsvine** [reply]
Thanks for the extra detail, @joelpeter94. I don't see anything glaringly slow/inefficient in your code, but I notice that it's not a complete/reproducible script. (E.g., `pagenumbers` is used here but not defined.) If you share a reproducible script and the PDF itself, perhaps I or someone else in the community can help pin down what's happening.

**Reply by joelpeter-c** [reply]
I am afraid I can share the pdf that I am trying to process (will check with the team though), but thank you for looking into it.

Much Appreciated and cheers.
